### ISSUE TITLE
Differential geometry

### ISSUE THREAD

**Issue by ocots** [control-toolbox/CTBase.jl#52]
@jbcaillau I have introduced more operations to compute controls, multipliers associated to state constraint of order 1, 2 and more, but also to compute singular arcs of any order.

For the moment, I have:

- Lift of a vector field: `Lift(X)` returns a Hamiltonian
- Lie derivatives: `Ad(X, f)` or `X⋅f` and `Ad(X, Y)` or `X⋅Y`
- Lie brackets: Lie(X, Y)

I will add of course Poisson brackets.

I need some validation for the syntax and possibilities. I present in the autonomous case but it works also in the non-autonomous case and in both vectorial and scalar cases.

## Lifts

From a `VectorField`:

```julia
julia> X = VectorField(x -> 2x)
julia> H = Lift(X)
julia> H(1, 1)
2
```

or in 2D:

```julia
julia> X = VectorField(x -> 2x)
julia> H = Lift(X)
julia> H([1, 2], [3, 4]) == 22
22
```

From a function it works also: `H = Lift(x -> 2x)`.

## Lie derivatives

```julia
julia> X = VectorField(x -> [x[2], -x[1]])
julia> f = x -> x[1]^2 + x[2]^2
julia> (X⋅f)([1, 2]) # or Ad(X, f)([1, 2])
0
```

Between two vector fields.

```julia
julia> X = VectorField(x -> [x[2], -x[1]])
julia> Y = VectorField(x -> [x[1], x[2]])
julia> (X⋅Y)([1, 2]) # or Ad(X, Y)([1, 2])
[2, -1]
```

You can also do:

```julia
julia> X = VectorField(x -> [x[2], -2x[1]])
julia> Y = VectorField(x -> [x[1], 3x[2]])
julia> f = x -> x[1]^2 + 2x[2]^2
julia> (X⋅Y⋅f)([1, 2])
```

## Lie brackets

```julia
julia> X = VectorField(x -> [x[2], -x[1]])
julia> Y = VectorField(x -> [x[2], -x[1]])
julia> Lie(X, Y)([1, 2])
[0, 0]
```

You can use the argument `order` to specify the bracket you want. For instance, let us consider

```julia
x = [1, 2, 3]
Γ = 2
γ = 1
δ = γ-Γ
F0 = VectorField(x -> [-Γ*x[1], -Γ*x[2], γ*(1-x[3])])
F1 = VectorField(x -> [0, -x[3], x[2]])
F2 = VectorField(x -> [x[3], 0, -x[1]])
```

Then,

```julia
F01 = Lie(F0, F1)
```

is equivalent to 

```julia
F01 = Lie(F0, F1, order=(1, 2))
# or
F01 = Lie(F0, F1, F2, order=(1, 2))
# or
F01 = Lie(F2, F0, F1, order=(2, 3))
```

We can also compute:

```julia
F120 = Lie(F0, F1, F2, order=(2,3,1))
```

If you want several Lie brackets of different lengths, you can do

```julia
F01 = Lie(F0, F1)
F012 = Lie(F01, F2)
```

or for others

```julia
F12, F121, F122 = Lie(F1, F2, orders=((1,2), (1,2,1), (1, 2, 2)))
```

Note that it is not so practical to give the orders like that so I have introduced the possibility to give labels:

```julia
x = [1, 2, 3]
Γ = 2
γ = 1
δ = γ-Γ
F0 = VectorField(x -> [-Γ*x[1], -Γ*x[2], γ*(1-x[3])], label=:F0)
F1 = VectorField(x -> [0, -x[3], x[2]], label=:F1)
F2 = VectorField(x -> [x[3], 0, -x[1]], label=:F2)
```

Then, you can do

```julia
F12, F121, F122 = Lie(F1, F2, orders=((:F1, :F2), (:F1, :F2, :F1), (:F1, :F2, :F2)))
```

**Comment by jbcaillau** [comment]
Very nice. Not sure to understand the "order" you introduced. (Seems to me that composing brackets in readable, and enough for applications.) One could alternatively define a macro so that
```julia
@Lie F10001
```
computed
```julia
Lie(F1, Lie(F0, ..., Lie(F0, F1)...)
```
(And similarly for Poisson brackets, *e.g.* Lie bracket and lifts.)

NB. A bit confused by your notations as
$$ [X,Y] = X\cdot Y-Y\cdot X = \mathrm{Ad}_X (Y) (\neq X\dot Y). $$

**Comment by ocots** [comment]
Ok I will make a pass on the notations. I used `Ad` since it was already existing.

The name `order` is not good I guess, I will find something else or remove it. The idea was to give all the vector fields to the function `Lie`, and then, composed them to make Lie brackets by giving something like `F10001` but I have chosen to give the indices of the vector fields in the function. 

```julia
F10 = Lie(F0, F1, F2, order=(2, 1))
F10001, F21, F201 = Lie(F0, F1, F2, orders=((2, 1, 1, 1, 2), (3, 2), (3, 1, 2))
```

`F1` is at the second place while `F0` is at the first one so to compute `F10` you give the "order" `(2, 1)`. I started from `1` since you can also write (you can switch places or have different names):

```julia
F10 = Lie(F2, F1, F0, order=(2, 3))
XY = Lie(X, Y, Z, order=(1, 2))
```

And I had introduced labels if giving the index was no so clear. 

I guess I can write you macro which calls my function. Do you prefer:

```julia
@Lie F10001
@Lie F21
@Lie F201
```

or

```julia
@Lie F10001, F21, F201
```

**Comment by ocots** [comment]
Some changes:

```julia
# lift
Lift(X)

# directional derivatives
Der(X, f)(x) = (X⋅f)(x) = Lie(X, f)(x) # = f'(x)⋅X(x)
Der(X, Y)(x) = (X⋅Y)(x) # = JY(x) X(x) 

# Lie brackets
Lie(X, Y) = ad(X, Y)= X⋅Y-Y⋅X # = [X, Y] 
Lie(X, Y, Z, sequence=(1, 3)) # = [X, Z]
Lie(X, Y, Z, sequence=(1, 3, 2)) # = [[X, Z], Y]
Lie(X, Y, Z, sequences=((1, 3), (1, 3, 2))) # = [X, Z], [[X, Z], Y]
Lie(X, Y, Z, sequence=(:X, :Z)) # = [X, Z], if the vector fields are have labels
@Lie F012 # F012 = [[F0, F1], F2]
```

But I don't manage to write a macro to compute several brackets in one line:

```julia
@Lie F01, F12, F101
```

Here is the macro for one bracket:

```julia
function _Lie(expr::Symbol)
    v = split(string(expr))[1]
    @assert v[1] == 'F'
    uni = sort(unique(v[2:end]))
    vfs = Tuple([Symbol(:F, n) for n ∈ uni])
    seq = [parse(Integer, c) for c ∈ v[2:end]]
    sequence = Tuple([findfirst(v->v==s, parse.(Integer, uni)) for s ∈ seq])
    code = quote
        $expr = Lie($(vfs...), sequence=$sequence)
    end
    return code
end

macro Lie(expr::Symbol)
    esc(_Lie(expr))
end
```

and this for several:

```julia
macro Lie(expr::Expr)
    @assert hasproperty(expr, :head)
    (@assert(s isa Symbol) for s ∈ expr.args)
    dump(expr)
    eval.([_Lie(s) for s ∈ expr.args])
end
```

But I want to remove `eval`. Any help?

**Comment by jbcaillau** [comment]
@ocots not sure about what you want to do but: regarding `eval`, you should be able to write your loop inside the generated code.

**Comment by ocots** [comment]
It was just to compute several brackets in one line with one call to `@Lie`. Not very useful.

Instead, I suggest this:

```julia
XY = @Lie [X, Y]
F = @Lie [[X, Y], [[X, Y], Z]]
```

To compute any Lie bracket.

**Comment by jbcaillau** [comment]
ouh, nice shot mr @ocots I did not even thought of macro to access [ ] brackets. Well done. Same for Poisson, BTW.

**Comment by ocots** [comment]
todo: remove `Der(X,Y)` and check `Lie(X, Y)` with the real intrinsic order 2 definition.

**Comment by jbcaillau** [comment]
Yes:
```julia
@test (@Lie [ X, Y ] ⋅ f)(x) == ((X⋅(Y⋅f))(x) - (Y⋅(X⋅f))(x)
```

**Comment by ocots** [comment]
Test passed:

```julia
        @testset "intrinsic definition" begin

            X = VectorField(x -> [x[2]^2, -2x[1]*x[2]])
            Y = VectorField(x -> [x[1]*(1+x[2]), 3x[2]^3])
            f = x -> x[1]^4 + 2x[2]^3
            x = [1, 2]

            # check real intrinsic order 2 definition of Lie bracket
            Test.@test ((@Lie [ X, Y ])⋅f)(x) == ((X⋅(Y⋅f))(x) - (Y⋅(X⋅f))(x))

        end
```

**Comment by ocots** [comment]
Ok what we have now:

```julia
# lift
H = Lift(X) # H(x, p) = p⋅X(x)

# directional derivatives
Lie(X, f)(x) = Der(X, f)(x) = (X⋅f)(x) # = f'(x)⋅X(x)

# Lie brackets
Lie(X, Y) = ad(X, Y) = @Lie [X, Y] # = [X, Y] 
F = @Lie [[X, Y], [[X, Y], Z]]
@Lie F012 # F012 = [[F0, F1], F2]
```

**Comment by ocots** [comment]
Still WIP.

**Comment by ocots** [comment]
Then, I have defined

https://github.com/control-toolbox/CTBase.jl/blob/e5b7a3ef2e4be85aacd6d573dde02223894da528/src/differential_geometry.jl#LL6C19-L6C19

which should be updated according to the new parameterization of functions, cf. [this comment](https://github.com/control-toolbox/CTBase.jl/issues/14#issuecomment-1540736199).

**Remark.** The Hamiltonian structure must be ` <: AbstractHamiltonian`.

Then, we need to had the Poisson bracket of two `AbstractHamiltonian`, according to

```julia
function Poisson(f, g)
    function fg(x, p)
        n = size(x, 1)
        ff = z -> f(z[1:n], z[n+1:2n])
        gg = z -> g(z[1:n], z[n+1:2n])
        df = ctgradient(ff, [ x ; p ])
        dg = ctgradient(gg, [ x ; p ])
        return df[n+1:2n]'*dg[1:n] - df[1:n]'*dg[n+1:2n]
    end
    return fg
end
```

See [Poisson bracket Wikipedia](https://en.wikipedia.org/wiki/Poisson_bracket).

**Remark.** If the two Hamiltonians are lifts of vector fields, then, there is a relation between the Poisson bracket of the Hamiltonians and the Lie bracket of the associated vector fields, cf. [ici](https://en.wikipedia.org/wiki/Poisson_bracket#A_result_on_conjugate_momenta) (not the best ref.).

**Comment by BaptisteCbl** [comment]
I have made some of the changes in this [branch](https://github.com/BaptisteCbl/CTBase.jl/tree/diff-geometry) of my fork but it is not finished, I need to add the macro for Poisson (the classic function works) and add tests for it.
Also Lift is broken for now because of the update from the original diff-geometry branch to the develop version of CTBase.
tl;dr: It is not finished yet on my side :disappointed: .

**Comment by jbcaillau** [comment]
@ocots @BaptisteCbl Hi! Curly brackets `{ , }` that are the standard notation for Poisson brackets are also available via macros, so it is possible to do the same as Olivier did for Lie ones:
```julia
julia> @dump { f, g }
Expr
  head: Symbol braces
  args: Array{Any}((2,))
    1: Symbol f
    2: Symbol g

julia> @Poisson { f, { f, g } }
```

**Comment by BaptisteCbl** [comment]
Hi, I have added the macro for Poisson and some tests. But before making the pull request I have a question.
I still don't know how to handle the variables of the non-fixed functions (`VectorField` and directional derivative for Lie and `AbstractHamiltonian` for Poisson), so there is no support for these non-fixed functions at the moment.
Does having a Poisson bracket of a variable dependent Hamiltonian have any meaning ? If so, does it have to be treated like the time variable of non autonomous functions ?

**Comment by ocots** [comment]
It has a meaning to compute Lie or Poisson brackets for functions depending on parameters such as a Variable. However we do not differentiate wrt the parameter. Maybe you can use `args...`.

**Comment by ocots** [comment]
However, for non autonomous vector fields or Hamiltonians, we should compute the partial derivative with respect to time in the Lie and Poisson brackets.

@jbcaillau what do you think?

**Comment by ocots** [comment]
@BaptisteCbl We can speak about that around 15:30 on discord if you want.

**Comment by BaptisteCbl** [comment]
Yes I would like to, thank you. I sent you a request on Discord but if you want to do it on other platform I'm okay too.

**Comment by ocots** [comment]
Discord is fine.

**Comment by ocots** [comment]
I need 10''. Do you find me on discord?

**Comment by BaptisteCbl** [comment]
Yes, I sent you a friend request and a message.

**Comment by jbcaillau** [comment]
@ocots @BaptisteCbl Hi guys, I would say yes: for time dependent functions, it makes sense to include time as an additional variable (for differentiation).

**Comment by ocots** [comment]
@BaptisteCbl I think you can abort the idea to add partial derivatives with respect to time in the Lie derivative, bracket and Poisson bracket. It seems difficult to have something consistent.we will see with use cases later.

**Comment by jbcaillau** [comment]
> @BaptisteCbl I think you can abort the idea to add partial derivatives with respect to time in the Lie derivative, bracket and Poisson bracket. It seems difficult to have something consistent.we will see with use cases later.

@ocots don't you have an example (irrigation thing...) with time dependent vector fields? it is good to start from a use case an check what kind of computation you need in this setting.

**Comment by ocots** [comment]
Actually, we need to differentiate in my use cases:

```julia
t -> g(t, x(t)) # state constraint
t -> H1(t, x(t), p(t)) # switching function
```

**Comment by ocots** [comment]
If we have:

```julia
ẋ(t) = F₀(t, x(t)) + u(t) F₁(t, x(t)) # the dynamics
H(t, x, p, u) = H₀(t, x, p) + u H₁(t, x, p) # the pseudo-Hamiltonian
g(t, x) ≤ 0 # the state constraint
```

We may need to compute:

```julia
t -> ∂ₜg(t, x(t)) + ∂ₓg(t, x(t))⋅ẋ(t)
t -> ∂ₜH₁(t, x(t), p(t)) + ∂ₓH₁(t, x(t), p(t))⋅ẋ(t) + ∂ₚH₁(t, x(t), p(t))⋅ẋ(t)
```

which is given for the moment by:

```julia
t -> ∂ₜg(t, x(t)) + Lie(F₁, g)(t, x(t))
t -> ∂ₜH₁(t, x(t), p(t)) + Poisson(H₀, H₁)(t, x(t), p(t))
```

But I have no example for the Lie bracket of two vector fields.

**Comment by jbcaillau** [comment]
@ocots Yes: adding the time as an additional state should work (there is indeed an associated dual variable, $p_t$). A good test is to write things for this example and check whether it remains user friendly, both from the mathematical and coding sides.

**Comment by ocots** [comment]
@BaptisteCbl Actually, it is ok. You can code the non-autonomous case with partial derivatives wrt time.

Let `X` being a vector field and `f` a differential function. We define

```julia
(X⋅f)(t, x) = ∂ₜf(t, x) + ∂ₓf(t, x)⋅X(t, x) # Lie derivative of f along X
[X, Y]⋅f = (X∘Y - Y∘X)⋅f # Lie bracket of X and Y
```

Then, 

```julia
([X, Y]⋅f)(t, x) = ∂ₓf(t, x)⋅( ∂ₜY(t, x) +  ∂ₓY(t, x)⋅X(t, x) - (∂ₜX(t, x) +  ∂ₓX(t, x)⋅Y(t, x)) )
```

which is ok since there is order 2. We thus define

```julia
(X ⅋ Y)(t, x) = ∂ₜY(t, x) +  ∂ₓY(t, x)⋅X(t, x) = Y'(t, x)⋅(1, X(t, x))
```

so that

```julia
[X, Y] = X ⅋ Y - Y ⅋ X
```

For the Poisson bracket, we denote by `F` the Hamiltonian vector field associated to `f`, a Hamiltonian now (I cannot write arrows). So:

```julia
F(t, z) = (∂ₚf(t, z), -∂ₓf(t, z)) # f(t, z) a non-autonomous Hamiltonian and F its Hamiltonian vector field
```

We define the Poisson bracket as:

```julia
{f, g} = F⋅g # the Lie derivative of g along F
```

Like this we have what we want since

```julia
{f, g}(t, z) = ∂ₜg(t, z) + ∂_z g(t, z)⋅F(t, z)
```

and let consider for instance that we want to differentiate wrt time `t -> g(t, z(t))` with `ż(t) = F(t, z(t))`. We get

```julia
dₜ g(t, z(t)) = ∂ₜg(t, z(t)) + ∂_z g(t, z(t))⋅ż(t) = {f, g}(t, z(t))
```

The only thing that changes is the following. Assume that

```julia
f(t, z) = p⋅X(t, x)
g(t, z) = p⋅Y(t, x)
```

Then,

```julia
{f, g}(t, z) = p⋅([X, Y](t, x) + ∂ₜX(t, x)) 
```

if I have not made a mistake.

**Comment by BaptisteCbl** [comment]
Ok, thanks ! I'll try to implement it in #85.

**Comment by jbcaillau** [comment]
@ocots Seems OK and in line with what you previously said, *i.e.* adding $t'(s)=1$ with $s$ a new time. Some thoughts:
- if there is a way to reuse previous (= autonomous) code, it has to be preferred... 
- ... provided everything remains user friendly (= use case compatible; you have a nice example to check)

I suppose that subtyping + dispatch *à la* `{<:TimeDependence}` might do the job.

**Comment by ocots** [comment]
We can reuse code since we can compute Poisson bracket from Lie derivative. 

Don't know if it is a good idea to use the autonomous case.

**Comment by jbcaillau** [comment]
hello mr @ocots!

**Comment by ocots** [comment]
👋

**Comment by jbcaillau** [comment]
mollo on coconut milk for breakfast, please 🤭

**Comment by ocots** [comment]
Et les langues, excellentes 😁

**Comment by ocots** [comment]
Et les mangues, excellentes 😁

**Comment by ocots** [comment]
Recap of what we will have:


Let `X` being a vector field and `f` a differentiable function. We define

```julia
(X⋅f)(t, x) = ∂ₓf(t, x)⋅X(t, x) # Lie derivative of f along X
[X, Y]⋅f = (X∘Y - Y∘X)⋅f # Lie bracket of X and Y
∂ₜ(f) = ∂ₜf
```

Then, 

```julia
([X, Y]⋅f)(t, x) = ∂ₓf(t, x)⋅(∂ₓY(t, x)⋅X(t, x) - (∂ₓX(t, x)⋅Y(t, x)))
```

We thus define

```julia
(X ⅋ Y)(t, x) =  ∂ₓY(t, x)⋅X(t, x)
```

so that

```julia
[X, Y] = X ⅋ Y - Y ⅋ X
```

For the Poisson bracket, we denote by `F` the Hamiltonian vector field associated to `f`, a Hamiltonian now (I cannot write arrows). So:

```julia
F(t, x, p) = (∂ₚf(t, x, p), -∂ₓf(t, x, p)) # f(t, x, p) a non-autonomous Hamiltonian and F its Hamiltonian vector field
```

We define the Poisson bracket as:

```julia
{f, g} = F⋅g # the Lie derivative of g along F
```

and let consider for instance that we want to differentiate wrt time `t -> g(t, x(t), p(t))` with `ż(t) = F(t, z(t))`, `z=(x,p)`. We will write

```julia
∂ₜ(g)(t, x(t), p(t)) + {f, g}(t, x(t), p(t))
```

**Comment by jbcaillau** [comment]
@ocots One quick remark on this:
![IMG_2141](https://github.com/control-toolbox/CTBase.jl/assets/62183989/05c89d43-1399-4a16-9cc5-285d55689123)

Regarding Lie derivatives, the idea is to differentiate $`f`$ along an integral curve of the nonautonomous vector field, $\dot{x}(t) = X(t, x(t))$, so
$$ \mathrm{d}/\mathrm{d}t (f(t, x(t))) = \partial_t f(t, x(t)) + \partial_x f(t, x(t)) \cdot X(t, x(t)). $$
Which is coherent with adding a new time $\mathrm{d}t/\mathrm{d}s = 1$ and extending the vector field.

**Comment by ocots** [comment]
Yes but in practice we do not want this to be automatic. I can explain on phone.

**Comment by ocots** [comment]
I think we can close this issue. We will talk about it :-)

**Comment by ocots** [comment]
If we have a constraint `c(t, x(t)) = 0` with a dynamics `ẋ(t) = F₀(t, x(t)) + u(t) F₁(t, x(t))`, then, to compute the control we will compute

```julia
dₜc(t, x(t)) = ∂ₜc(t, x(t)) + ∂ₓc(t, x(t))⋅ẋ(t) 
  = ∂ₜc(t, x(t)) + ∂ₓc(t, x(t))⋅(F₀(t, x(t)) + u(t) F₁(t, x(t))) = 0
```

that is 

```julia
0 = ∂ₜc + F₀⋅c + u F₁⋅c
```

so we do not want to have `(X⋅f)(t, x) = ∂ₜf(t, x) + ∂ₓf(t, x)⋅X(t, x)` since otherwise we would have twice `∂ₜc`.

**Comment by jbcaillau** [comment]
I see your point. We discuss it this Friday in Marseille. Keeping the issue open.

**Comment by jbcaillau** [comment]
@ocots @BaptisteCbl One more thing: looks a bit redundant to write
```julia
@Poisson { H0, H1 }
@Lie [ F0, F1 ]
...
```
Maybe better to use a single macro (`@Lie`?) with different brackets / symbols for different operations:
```julia
@Lie { H0, H1 }
@Lie [ F0, F1 ]
@Lie F ⋅ g
```

**Comment by j-l-s** [comment]
Coherence is the master word for me and if brackets/curly_brackets
are not significient, then keep only one option:

@Poisson {H0, H1}  + @Lie {F0, F1}

or

@Poisson [H0, H1]  + @Lie [ F0, F1]

or

@Poisson (H0, H1)  + @Lie (F0, F1)

or

@Poisson H0, H1  + @Lie F0, F1

but do not make it confusing for end users…..


(why not : @Poisson { [ (H0), (H1)] }  and @Lie F0 / F1 ????



> On 7 Jun 2023, at 23:20, Jean-Baptiste Caillau ***@***.***> wrote:
> 
> 
> @ocots @BaptisteCbl One more thing: looks a bit redundant to write
> @Poisson { H0, H1 }
> @Lie [ F0, F1 ]
> ...
> Maybe better to use a single macro ***@***.***?) with different brackets / symbols for different operations:
> @Lie { H0, H1 }
> @Lie [ F0, F1 ]
> @Lie F ⋅ g
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>

—
Jean-Luc Szpyrka

**Comment by jbcaillau** [comment]
Hello nightly @j-l-s ! Actually, it is exactly the point: `{ , }` and `[ , ]` are distinct notations for distinct computations. The first one is for Poisson brackets of (scalar valued) functions, while the second is for Lie brackets of vector fields. (And the third notation `F ⋅ g` is for yet another computation...) Prefixing with a macro is just a convenient way (👍🏽 @ocots) to redefine brackets (which is not allowed otherwise). But then, only one macro `@foo` is enough. And since all three computations are, in some general sense, "Lie derivatives", it makes sense to call the macro `@Lie` (instead of `@foo`).

**Comment by jbcaillau** [comment]
@BaptisteCbl @ocots are you OK with the proposal below?
```julia
@L { f, g } # for real f(x, p), x and p vectors of same dim; calls Poisson(f, g)
@L [ X, Y ] # for vector X(x), x vector; calls Lie(X, Y)
X ⋅ f # for vector X(x), x vector, scalar f(x); calls Der(X, f)
```
where `L` stand for Lie or whatever you want (or replace it by sth else). Note that no macro is needed for `⋅`

See also: #101

**Comment by ocots** [comment]
I have a question we have to answer before making any update. For the moment we can compute the Lie derivative of a `Function` along a `VectorField`. The time dependence of the function is assumed to be the same as for the vector field. Idem for variable dependence. So in this case there is no problem.

If we allow to compute the Lie derivative of a `Function` along a vector field which is also a `Function`, we have to assume the time and variable dependencies. Again this is not a problem. We can assume that the functions are autonomous and have no variables. 

But, if we use the dot operation, we cannot compute Lie derivatives with two functions if the dependencies are not the default ones. Is this a problem?

When computing Lie and Poisson brackets, there is no problem since we can add options for the dependencies. Note that this haven't been implemented.

**Comment by ocots** [comment]
We have converged to:

```julia
@Lie { f, g } # Poisson bracket
@Lie [X, Y] # Lie bracket
X ⋅ f # Lie derivative
```

If the data have no specific types, then we assume the functions are autonomous and variable independent.

**Comment by ocots** [comment]
Some propositions, just to write them

```julia
@L { f, g }
@DG { f, g }
@Geo { f, g }
@Geometry { f, g }
@DiffGeo { f, g }
```

**Comment by jbcaillau** [comment]
> Some propositions, just to write them
> 
> ```julia
> @L { f, g }
> @DG { f, g }
> @Geo { f, g }
> @Geometry { f, g }
> @DiffGeo { f, g }
> ```

May be `@Lie` is my favourite (short but explicit).

**Comment by jbcaillau** [comment]
@ocots hi again, can you please comment on what you actually merged?
